plugins/database/mongodb/mongodb.go

```diff
@@ -135,17 +135,17 @@ func (m *MongoDB) CreateUser(ctx context.Context, statements dbplugin.Statements
 		Roles:    mongoCS.Roles.toStandardRolesArray(),
 	}
 
-	err = session.DB(mongoCS.DB).Run(createUserCmd, nil)
+	result := client.Database(mongoCS.DB).RunCommand(ctx, createUserCmd, nil)
 	switch {
-	case err == nil:
-	case err == io.EOF, strings.Contains(err.Error(), "EOF"):
-		// Call getConnection to reset and retry query if we get an EOF error on first attempt.
-		session, err := m.getConnection(ctx)
+	case result.Err() == nil:
+	case result.Err() == io.EOF, strings.Contains(result.Err().Error(), "EOF"):
```

Comment:
`ErrNoDocuments` is returned from operations like `FindOne` when no documents were returned by the server. It should never be returned from `RunCommand`, because the server will at the very least respond with `{ok: 0}` or `{ok: 1}` depending on whether or not the operation succeeded.

Can you explain what kinds of errors this code is targeting for retries? Note that versions 1.1.0 of the driver do support retryable reads per the MongoDB driver specification (https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst), but this is not being done for `RunCommand` and `RunCommandCursor`. There is a ticket to support this feature for these methods (https://jira.mongodb.org/browse/GODRIVER-1252) so if the retry logic here is messy, feel free to comment on that ticket and we'll look at its priority again during our next meeting.

Reply:
See https://github.com/hashicorp/vault/pull/8140#discussion_r366770775